### PR TITLE
    Refactor the authentication system

### PR DESCRIPTION
**Pull request by RedGuy12**
### Resolves

Resolves #

### Changes

 **auth api changes**
- change `sendResponse` from a 3rd parameter to `this.sendResponse`. requires we use **non-arrow functions** for the request handlers. also use non-arrow functions for `getData` just to be consistent
- change `pages` from an array of objects to a single object.
- getData: instead of returning a falsy value on error, return an object with an `error` property.
- consistent `return` statements
- use JWT :eyes: :eyes: :eyes: incl. remove `getData`, `rawData`, and update types
- replit auth: send more info on errors

**core changes**
- use the client name instead of the page name in `send_data`
- reorder handlers
- check url validity on `/auth` as well as on the allow page
- update parameters of the `nonce` cookie
- loop over each handler instead of each method
- consistent `return` statements
- support `app.all`
- better code overall
- remove `/backend/send_data` (uneeded with jwt)
- remove `AuthDatabase`

**small unrelated changes**
- update the options for the `language` cookie
- reorder some middlewares
- format some html comments
- change comment in lighthous workflow
- bump eslint and eslint plugins/configs

### Reason for changes

cleaner code, security

### Agree to our policies

-   [x] I agree to follow the [Code of Conduct](https://github.com/onedotprojects/auth/blob/issues/.github/CODE_OF_CONDUCT.md).
-   [x] I agree to follow the [Contributing Guidelines](https://github.com/onedotprojects/auth/blob/issues/.github/CONTRIBUTING.md).

### todo
- use `all`
- docs
- add env vars to opeNode
- remove noData strings

### CODE REVIEW

**Comment by wgyt** [comment]
why do you hate arrow functions

**Comment by RedGuy12** [comment]
> - change `sendResponse` from a 3rd parameter to `this.sendResponse`. requires we use **non-arrow functions** for the request handlers. also use non-arrow functions for `getData` just to be consistent

> why do you hate arrow functions

File: lib/localization.js [hunk 1]
```diff
@@ -426,8 +426,11 @@ export default function localization(request, response, next) {
 
 	response.cookie("languages", languages.join("|"), {
 		expires,
+		httpOnly: false,
 		maxAge: 31536000000,
-		sameSite: false,
+		sameSite: "none",
+		secure: true,
+		signed: false,
```
Comment (wgyt) [inline lib/localization.js:433]: why do we need these headers

File: lib/localization.js [hunk 2]
```diff
@@ -443,7 +446,7 @@ export default function localization(request, response, next) {
 	 * Override `res.render` to ensure `message` is always available.
 	 *
 	 * @param {string} view - The file to render.
-	 * @param {{ [key: string]: any } | ((error: Error, str: string) => undefined)} [variablesOrCallback]
+	 * @param {{ [key: string]: any } | ((error: Error, str: string) => void)} [variablesOrCallback]
```
Comment (wgyt) [inline lib/localization.js:449]: what is the difference between undefined and void

File: lib/mongoose.js [hunk 3]
```diff
@@ -18,25 +18,7 @@ export const database = mongoose.connection;
...
-			token: {
-				match: /^[\da-z]{10}$/,
-				required: true,
-				type: String,
-				unique: true,
-			},
-		}),
-	),
-	EmailDatabase = mongoose.model(
+export const EmailDatabase = mongoose.model(
```
Comment (wgyt) [inline lib/mongoose.js:21]: why remove AuthDatabase

File: src/auth/discord/index.js [hunk 4]
```diff
@@ -7,46 +7,41 @@ dotenv.config();
...
-
-		user.connections = await fetch("https://discord.com/api/users/@me/connections", {
-			headers: {
-				authorization: `${tokens.token_type} ${tokens.access_token}`,
-			},
-		}).then((result) => result.json());
-
-		return user;
-	},
-
```
Comment (wgyt) [inline src/auth/discord/index.js:37]: why remove getData

File: src/auth/discord/index.js [hunk 5]
```diff
@@ -7,46 +7,41 @@ dotenv.config();
...
+					headers: {
+						authorization: `${tokens.token_type} ${tokens.access_token}`,
+					},
+				}).then((result) => result.json());
+
+				return this.sendResponse(user, `${request.query.state}`);
+			},
 		},
-	],
+	},
```
Comment (wgyt) [inline src/auth/discord/index.js:44]: why is this not an array anymore

File: src/auth/email/index.js [hunk 6]
```diff
@@ -53,7 +51,7 @@ const client = {
 
 					await EmailDatabase.deleteOne({ code: request.body.code }).exec();
 
-					return sendResponse({ email }, `${request.query?.nonce}`);
+					return this.sendResponse({ email }, `${request.query?.nonce}`);
```
Comment (wgyt) [inline src/auth/email/index.js:54]: why is sendResponse in this?

File: src/auth/email/index.js [hunk 7]
```diff
@@ -106,9 +104,7 @@ const client = {
 				return response.status(400);
 			},
 		},
-	],
-
-	rawData: true,
```
Comment (wgyt) [inline src/auth/email/index.js:111]: why remove rawData?

File: src/auth/replit/index.js [hunk 8]
```diff
@@ -7,44 +7,54 @@ import { logError } from "../../errors/index.js";
...
 	name: "Replit",
 
-	pages: [
-		{
-			backendPage: "../__replauth",
-
-			get: (request, _, sendResponse) =>
-				sendResponse(`${request.query.token}`, `${request.query.redirect}`),
+	pages: {
+		"/__replauth": {
```
Comment (wgyt) [inline src/auth/replit/index.js:15]: ../ isn't needed anymore ? 
./ equals /auth/ ?

File: src/partials/footer.html [hunk 9]
```diff
@@ -125,7 +125,7 @@
 					>
 				</li>
 				<li>
-					<!--TODO: make a privacy page-->
+					<!-- TODO: make a privacy page -->
```
Comment (wgyt) [inline src/partials/footer.html:128]: just use the 1dot one
https://1dot.cf/legal

File: lib/localization.js (hunk 2 above)
Comment (RedGuy12) [reply lib/localization.js:449]: void: doesn't return anything
undefined: returns undefined
any: this is anything
unknown: don't try to do anything with this

Comment (RedGuy12) [reply lib/localization.js:449]: probably should be unknown instead

File: lib/localization.js (hunk 1 above)
Comment (RedGuy12) [reply lib/localization.js:433]: they are important

plus they should improve the lighthouse score a tiny bit

File: lib/mongoose.js (hunk 3 above)
Comment (RedGuy12) [reply lib/mongoose.js:21]: because we use jwt instead now

File: src/auth/discord/index.js (hunk 4 above)
Comment (RedGuy12) [reply src/auth/discord/index.js:37]: it's moved into the page handler now, that's how jwt works

File: src/auth/discord/index.js (hunk 5 above)
Comment (RedGuy12) [reply src/auth/discord/index.js:44]: so the code in `src/auth/index.js` can be simpler

File: src/auth/email/index.js (hunk 7 above)
Comment (RedGuy12) [reply src/auth/email/index.js:111]: same reason why i removed `getData`

File: lib/localization.js (hunk 2 above)
Comment (wgyt) [reply lib/localization.js:449]: ok, got it

File: lib/localization.js (hunk 1 above)
Comment (wgyt) [reply lib/localization.js:433]: ah ok

File: src/auth/replit/index.js (hunk 8 above)
Comment (RedGuy12) [reply src/auth/replit/index.js:15]: `../` means go back one level
`./` (or if it's omnitted) means stay at this level
`/` means go to the root

File: src/partials/footer.html (hunk 9 above)
Comment (RedGuy12) [reply src/partials/footer.html:128]: i already link to that

i want to make a separate privacy page for oneauth

we can discuss in discord about this, not here, it's not part of this pr

File: src/auth/email/index.js (hunk 6 above)
Comment (RedGuy12) [reply src/auth/email/index.js:54]: ...why not? :P

i just wanted the call signature to be the same as express `app.get`

**Comment by RedGuy12** [comment]
will merge after i write docs

File: src/partials/head.html [hunk 10]
```diff
@@ -1,17 +1,19 @@
 <!-- Google Tag Manager -->
 <script>
-	"use strict";
-
-	(function googleTagManager() {
+	"use strict"(function googleTagManager() {
```
Comment (RedGuy12) [inline src/partials/head.html:3]: oops

**Pull request merged**